File: Sorting/Bubble/BubbleSort.py
# User-defined function for Bubble Sort
def bubble_sort(arr):
    n = len(arr)
    for i in range(n - 1):
        swapped = False
        for j in range(n - i - 1):
            if arr[j] > arr[j + 1]:
                # Swap arr[j] and arr[j+1]
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                swapped = True
        # If no swapping happened, array is already sorted
        if not swapped:
            break

# User-defined function to print an array
def print_array(arr):
    print(" ".join(map(str, arr)))

# Taking input from user
n = int(input("Enter the number of elements: "))
arr = list(map(int, input(f"Enter {n} elements: ").split()))

# Ensure the correct number of elements are entered
while len(arr) != n:
    print(f"Error: You must enter exactly {n} elements. Try again.")
    arr = list(map(int, input(f"Enter {n} elements: ").split()))

print("Unsorted array:")
print_array(arr)

# Calling the user-defined function for Bubble Sort
bubble_sort(arr)

print("Sorted array:")
print_array(arr)

File: ARRAY/Sorting/Bubble/BubbleSort.c
#include <stdio.h>

// User-defined function for Bubble Sort
void bubbleSort(int arr[], int n) {
    int swapped;
    for (int i = 0; i < n - 1; i++) {
        swapped = 0;
        for (int j = 0; j < n - i - 1; j++) {
            if (arr[j] > arr[j + 1]) {
                // Swap arr[j] and arr[j+1]
                int temp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = temp;
                swapped = 1;
            }
        }
        // If no swapping happened, array is already sorted
        if (!swapped) break;
    }
}

// User-defined function to print an array
void printArray(int arr[], int n) {
    for (int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");
}

int main() {
    int n;

    // Taking input from user
    printf("Enter the number of elements: ");
    scanf("%d", &n);

    int arr[n]; // Declare an array of size n

    printf("Enter %d elements:\n", n);
    for (int i = 0; i < n; i++) {
        scanf("%d", &arr[i]);
    }

    printf("Unsorted array:\n");
    printArray(arr, n);

    // Calling the user-defined function for Bubble Sort
    bubbleSort(arr, n);

    printf("Sorted array:\n");
    printArray(arr, n);

    return 0;
}
